embassy-stm32/src/sdmmc/mod.rs

```diff
@@ -1478,3 +1478,39 @@ foreach_peripheral!(
         impl Instance for peripherals::$inst {}
     };
 );
+
+impl<'d, T: Instance, Dma: SdmmcDma<T> + 'd> block_device_driver::BlockDevice<512> for Sdmmc<'d, T, Dma> {
+    type Error = Error;
+
+    async fn read(&mut self, mut block_address: u32, buf: &mut [[u8; 512]]) -> Result<(), Self::Error> {
+        assert!(&buf[0] as *const _ as u32 % 4 == 0);
```

Comment:
if the trait allows unaligned buffers, this code should work with unaligned buffers IMO. Otherwise you get asserts at runtime, have to add hacks to trait users so they align the buffer even if it's not compile-time checked, and bad compatibility across impls.

(There's a similar problem with nRF's flash: it also asserts at runtime the buffer is 4-aligned when the embedded-storage traits don't require it, and it has caused quite a bit of pain. I see it as a bad mistake that I regret and should fix someday...)

Reply:
Hmm. I wanted to keep the block device trait as simple as possible, but maybe adding some kind of alignment constraint is worthwhile.

I've already explored this with https://docs.rs/block-device-adapters/latest/block_device_adapters/struct.BufStream.html which you'll need anyways to use embedded-fatfs, so in _this_ case the assert is more of fail safe. However other fs impls may choose to use the block device directly.

I think I could just clean up and use [AlignedBuffer](https://github.com/MabezDev/embedded-fatfs/blob/11bd13515886c539922087903a7d35657e8b9a3f/block-device-adapters/src/buf_stream.rs#L271-L277), but I'd need to explore if can still be cast from byte slice to block slice and vice versa :thinking:. Generic alignment can be a bit tricky in Rust.

Reply:
I think that buffer alignment is going to be common, I looked at the sdmmc peripheral for the esp's and for DMA usage they require word-aligned buffers too, I'm going to look into adding some kind of alignment constraint to the trait and report back.